### ISSUE TITLE
2.5.8 "DELETE ALL RULES" REST API doesn't delete the INTEGRITY rule.

### ISSUE THREAD

**Issue by CesuCiso** [Apicurio/apicurio-registry#4226]
Version 2.5.8

**First step** 
```
 url = schemaRegistryURL + "/groups/" + encodeValue(appName) + "/artifacts/" + encodeValue(artifactID) + "/rules";
 System.out.println("Delete all existing rules for schema=" + appName + "/" + artifactID);
 HttpDelete  httpDelete = new HttpDelete(url);

try  {   
    try (CloseableHttpResponse response2 = httpclient.execute(httpDelete)) {
        HttpEntity entity2 = response2.getEntity();
        EntityUtils.consume(entity2);
        switch (response2.getCode()) {
            case (204):
                System.out.println("OK.");
                break;
                 ...
```
=> return OK

**Second step immediatly after first step**

```
url = schemaRegistryURL + "/groups/" + encodeValue(appName) + "/artifacts/" + encodeValue(artifactID) + "/rules/INTEGRITY";
System.out.println("Verif INTEGRITY rule is deleted");
httpGet = new HttpGet(url);
try (CloseableHttpResponse response2 = httpclient.execute(httpGet)) {
    HttpEntity entity2 = response2.getEntity();
    EntityUtils.consume(entity2);
    switch (response2.getCode()) {
        case (200):
            System.err.println("INTEGRITY rule is not deleted");
            System.exit(2);
            break;
        case (400):
            System.err.println("Error request");
            System.err.println("Invalid rule type");
            System.exit(2);
            break;
        case (404):
            System.out.println("INTEGRITY rule is deleted");
            break;
        default: 
            System.err.println ("Error request");
            System.err.println("Unrecognized response code: " + response2.getCode() + "/" + response2.getReasonPhrase());
            System.exit(2);
    }
}
```
=> return INTEGRITY rule is not deleted.

**Comment by EricWittmann** [comment]
I cannot reproduce this in a test.  Please see this PR: https://github.com/Apicurio/apicurio-registry/pull/4242

The test passes for me locally using SQL storage.  Let's see if it passes in CI/CD for all storage variants, but I suspect it will.

If you can create a simple Java project that reproduces this problem that I can run, I will debug further.

**Comment by CesuCiso** [comment]
You just have to comment this part, to verify that the integrity rule is not deleted:

 // Apicurio bug: integrity rule is not deleted by previous call: delete specifically integrity rule
 url = schemaRegistryURL + "/groups/" + encodeValue(appName) + "/artifacts/" + encodeValue(artifactID) + "/rules/INTEGRITY";
 System.out.println("Apicurio Bug: Delete INTEGRITY rule for schema=" + appName + "/" + artifactID);
 itemDelete(httpclient, url, "rule");

**Comment by EricWittmann** [comment]
Thanks!  I'll check out your code as soon as I can.  What you provided looks perfectly fine for a reproducer.  I would have also accepted an attached .java file or a link to a GitHub repository.  :)  

Now we need to have a contest to see who is an older Java dev.  LOL

**Comment by EricWittmann** [comment]
OK I have spent some time with this, but I still cannot reproduce the reported problem using your provided code.  Perhaps I'm not doing it the same way you are, since your reproducer is basically a CLI that can be executed a number of different ways, with different .avsc file inputs.  So without further instructions from you, I tried to figure out what you might have been doing.  To that end, I have modified the code a little bit and uploaded it all in a reproducible manner here:

https://github.com/EricWittmann/apicurio-registry-delete-all-rules

If you clone that repository you can follow the instructions in the readme and it will result in a successful run.

If you can make it fail, please let me know how - or better yet submit a pull request with the changes needed.

**Comment by EricWittmann** [comment]
I realized that you mentioned using the KafkaSQL storage, which I didn't test.  It's possible that there is a bug only in that storage - I'll try to find some time to test that out tomorrow.

**Comment by EricWittmann** [comment]
Working on a fix for this, the PR will be here:

https://github.com/Apicurio/apicurio-registry/pull/4282

The PR is already opened to ensure that the test I wrote for this problem actually fails (successfully reproduces the bug).  Now that that has been confirmed I will work on a fix.

**Comment by EricWittmann** [comment]
I have a fix and am awaiting review before merging.  @CesuCiso thanks for the bug report.

**Comment by EricWittmann** [comment]
Fixed in https://github.com/Apicurio/apicurio-registry/pull/4282.

**Comment by CesuCiso** [comment]
Hello Eric.
Are you sure the patch works?
I just installed 2.5.11
I verified that Apicurio 2.5.11 is running (/apis/registry/v2/system/info => 2.5.11.Final)

I call http DELETE on /groups/MyGroup/artifacts/artifactID/rules. I get a response code 204.

Then I call http POST on /groups/MyGroup/artifacts/MyArtifactID/rules. (validity full). I get a response code 200

Then I call http POST on /groups/MyGroup/artifacts/MyArtifactID/rules. (compatibility backward_transitive). I get a response code 200

Then I call http POST on /groups/MyGroup/artifacts/MyArtifactID/rules. (integrity full). I get a response code 409.

{
"message": "A rule named 'INTEGRITY' already exists.",
"error_code": 409,
"detail": "RuleAlreadyExistsException: A rule named 'INTEGRITY' already exists.",
"name": "RuleAlreadyExistsException"
}
